### PR TITLE
Drop Ruby < 2.7 and deal with kwargs

### PR DESCRIPTION
**Pull request by ekohl**
Ruby 3 has started to separate args and kwargs.

### CODE REVIEW

**Comment by ekohl** [comment]
This fails on Ruby < 2.7. We can either use some complicated method to support both, or move forward and drop it. Today Foreman is the main consumer and that only cares about Ruby 2.7+. 2.6 and older are EOL anyway.

File: lib/ruby_parser_string_io_patch.diff
Comment (ekohl) [inline lib/ruby_parser_string_io_patch.diff:2]: Now I'm reading through this and wonder why this file is even here. I think it can be dropped. It was added in eb79570c2794415cc3d5d960cccb7ccda08f136f.

**Review by ares — CHANGES_REQUESTED** [review]
I've tested this with Foreman and it shows there's some change in the behavior. The kickstart template no longer works, complaining about `kickstart_ifcfg_get_identifier_names` snippet calling `inheriting_mac` on `nil`. This method is called on `@interface` which is supposed to be a variable holding the interface object. Reading the code, it looks like we can no longer pass variables through the snippet macro.

It seems like the `Safemode::Box#eval` no longer supports second argument of allowed variables (assigns). While this PR does not touch the code, I guess the kwargs somehow impacts that. However this code seems to work fine

```ruby
require 'safemode'
require 'erb'

erb_code = %q{<%= @a %>}
box = Safemode::Box.new

puts "Doing the ERB code in safe mode\n-----"
puts box.eval(ERB.new(erb_code).src, { :a => 'b' })
```

I couldn't find the actual cause, but it's quite simple to reproduce with Foreman. Just create a snippet with `<%= @a %>` and then call it from the template `<%= snippet('name', variables: { :a => 'b' }) %>`

File: lib/safemode/scope.rb
Comment (ares) [inline lib/safemode/scope.rb:37]: is this still a TODO? should we keep it or resolve it before the merge?

**Comment by ares** [comment]
And I can confirm, it's the last commit that caused the change in the behavior.

**Comment by ekohl** [comment]
> I've tested this with Foreman and it shows there's some change in the behavior.

Since there is a difference in Ruby versions, did you test it with Ruby 2.7 or 3.0?

> And I can confirm, it's the last commit that caused the change in the behavior.

Any thoughts on already merging the first 3 commits?

**Comment by ares** [comment]
It's 2.7.6 available on CentOS Stream 8.

Happy to merge the first three commits if you open them as a separate PR

**Comment by ekohl** [comment]
I've pushed a fix that locally made Foreman's `rake test TEST=test/controllers/api/v2/report_templates_controller_test.rb:229` pass. However, I'm still running more tests and I also see `undefined method '#name' for NilClass::Jail`.

File: test/test_safemode_parser.rb
Comment (ekohl) [inline test/test_safemode_parser.rb:30]: I tracked it down to this bit. Reading https://github.com/seattlerb/ruby2ruby/pull/55#issuecomment-1100342149 I'm afraid we may have to dive into ruby2ruby to properly deal with this.

Comment (ekohl) [reply test/test_safemode_parser.rb:30]: I should note that ruby2ruby 2.5.0 does add support for kwargs, so we can at least do it correctly.

File: test/test_safemode_parser.rb
Comment (ofedoren) [reply test/test_safemode_parser.rb:42]: That indeed looks awkward..

**Review by ofedoren — COMMENTED** [review]
Hmm... As of now it looks sane. I've locally run Foreman tests with this PR and they are green (Ruby 2.7.6).

Although, I'd prefer to extract safe call operator from this PR (I know there is one already) and merge that first.

I'm going to run the tests on Ruby 3 and I'll try to do some testing by hands. If that passes as well, I'd be OK with merging even some tests look awkward :)

I'd highly appreciate some @adamruzicka's eyes on this as well.

**Comment by ofedoren** [comment]
Update: I've run the set of Foreman's tests on Ruby 3.0.4 and with https://github.com/theforeman/foreman/pull/9871 it was green. At least on my machine. I've tried to run GHA with that (https://github.com/ofedoren/foreman/pull/9), but I forgot that this lib is still pinned to Ruby < 3.0.

@ekohl, since we're trying to make this lib compatible with Ruby 3, could we update the pinned version? It'll help with automated tests. Which reminds me, maybe we could update ci.yml here to run the tests on Ruby 3.0?

Looking into manual testing now.

**Review by ofedoren — APPROVED** [review]
Thanks, all, this seems to be ready. I've also included Ruby 3.0 CI and the tests are green. Let's get this in.

**Pull request merged**
